plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,456 @@
+};
+use crate::plonk::config::Hasher;
+use crate::plonk::config::{AlgebraicHasher, GenericConfig};
+use crate::plonk::proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget};
+use crate::recursion::conditional_recursive_verifier::dummy_proof;
+
+pub struct CyclicRecursionData<
+    'a,
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    proof: &'a Option<ProofWithPublicInputs<F, C, D>>,
+    verifier_data: &'a VerifierOnlyCircuitData<C, D>,
+    common_data: &'a CommonCircuitData<F, D>,
+}
+
+pub struct CyclicRecursionTarget<const D: usize> {
+    pub proof: ProofWithPublicInputsTarget<D>,
+    pub verifier_data: VerifierCircuitTarget,
+    pub dummy_proof: ProofWithPublicInputsTarget<D>,
+    pub dummy_verifier_data: VerifierCircuitTarget,
+    pub base_case: BoolTarget,
+}
+
+impl<C: GenericConfig<D>, const D: usize> VerifierOnlyCircuitData<C, D> {
+    fn from_slice(slice: &[C::F], common_data: &CommonCircuitData<C::F, D>) -> Result<Self>
+    where
+        C::Hasher: AlgebraicHasher<C::F>,
+    {
+        // The structure of the public inputs is `[..., circuit_digest, constants_sigmas_cap]`.
+        let cap_len = common_data.config.fri_config.num_cap_elements();
+        let len = slice.len();
+        ensure!(len >= 4 + 4 * cap_len, "Not enough public inputs");
+        let constants_sigmas_cap = MerkleCap(
+            (0..cap_len)
+                .map(|i| HashOut {
+                    elements: std::array::from_fn(|j| slice[len - 4 * (cap_len - i) + j]),
+                })
+                .collect(),
+        );
+        let circuit_digest =
+            HashOut::from_partial(&slice[len - 4 - 4 * cap_len..len - 4 * cap_len]);
+
+        Ok(Self {
+            circuit_digest,
+            constants_sigmas_cap,
+        })
+    }
+}
+
+impl VerifierCircuitTarget {
+    fn from_slice<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+        slice: &[Target],
+        common_data: &CommonCircuitData<F, D>,
+    ) -> Result<Self> {
+        let cap_len = common_data.config.fri_config.num_cap_elements();
+        let len = slice.len();
+        ensure!(len >= 4 + 4 * cap_len, "Not enough public inputs");
+        let constants_sigmas_cap = MerkleCapTarget(
+            (0..cap_len)
+                .map(|i| HashOutTarget {
+                    elements: std::array::from_fn(|j| slice[len - 4 * (cap_len - i) + j]),
+                })
+                .collect(),
+        );
+        let circuit_digest = HashOutTarget {
+            elements: std::array::from_fn(|i| slice[len - 4 - 4 * cap_len + i]),
+        };
+
+        Ok(Self {
+            circuit_digest,
+            constants_sigmas_cap,
+        })
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> CircuitBuilder<F, D> {
+    /// Add verifier data and register it as public inputs.
+    /// WARNING: Do not register any public input after calling this!
```

Comment:
Seems fine for now but is this something we could relax later? Maybe this function could return the start index of cyclic recursion PIs, and that start index could then be passed to `cyclic_recursion` and to `from_slice`.

Reply:
Yes it should be easy to relax if needed. I thought it wouldn't be too annoying like that since the recursion gadget could typically be called at the end of the circuit, but if it becomes annoying this can be changed as you suggested.